### PR TITLE
Update Undo recipe

### PR DESCRIPTION
**Pull request by HansDP**
Modified the Undo recipe to use a race condition
Related issue: #647

### CODE REVIEW

**Comment by HansDP** [comment]
I did some basic smoke testing on the code snippet. It runs and everything seems fine.

PS: this is the first time I created a pull request on github. If I did anything wrong, just let me know

File: docs/recipes/README.md [hunk 1]
```diff
@@ -154,43 +154,54 @@ creates a higher order reducer to do most of the heavy lifting for the developer
...
 function* main() {
   while (true) {
-    // listen for every `ARCHIVE_THREAD` action in a non-blocking manner.
-    const onArchiveTask = yield takeEvery(ARCHIVE_THREAD, onArchive)
-    // wait for the user to activate the undo action;
-    yield take(UNDO)
-    // and then cancel the fetch task
-    yield cancel(onArchiveTask)
+    // wait for an ARCHIVE_THREAD to happen
+    const action = yield take(`ARCHIVE_THREAD`)
```
Comment (Andarist) [inline docs/recipes/README.md]: I wouldnt use `` syntax here, as you are not interpolating anything. '' would suffice

File: docs/recipes/README.md [hunk 2]
```diff
@@ -154,43 +154,54 @@ creates a higher order reducer to do most of the heavy lifting for the developer
...
+    // use spawn to execute onArchive in a non-blocking fashion (which also
```
Comment (Andarist) [inline docs/recipes/README.md]: you have the opening bracket here in the comment, but no closing one

**Comment by Andarist** [comment]
It looks rly great, I just aint sure if introducing `spawn` here is necessary. It convolutes the example a little bit. Anyway - `fork` vs `spawn` struggle is geting bigger and bigger in the community. Gonna speak about it with @yelouafi 

I've put 2 really minor comments, if you decide to correct them, please do an ammend commit (or squash them) + rebase with the current master - it will help keeping the history clean.

**Comment by HansDP** [comment]
I agree about the `spawn` vs `fork` comment.
But I'm not sure how to handle cancellation in `onArchive` when the `main` saga gets cancelled.

Just 'ignore' this (to keep the example clear) is an option, but in production code, i guess there must be some flags to know whether or not the undo ui should be hidden and whether or not the optimistic update should be reverted.

Something like

```javascript
import { take, put, call, fork, race } from 'redux-saga/effects'
import { delay } from 'redux-saga'
import { updateThreadApi, actions } from 'somewhere'

function* onArchive(action) {

  let needsHideUndo = false
  let needsRevertOptimisticUpdate = false

  const { threadId } = action
  const undoId = `UNDO_ARCHIVE_${threadId}`

  const thread = { id: threadId, archived: true }

  try {

    // show undo UI element, and provide a key to communicate
    yield put(actions.showUndo(undoId))
    needsHideUndo = true

    // optimistically mark the thread as `archived`
    yield put(actions.updateThread(thread))
    needsRevertOptimisticUpdate = true

    // allow the user 5 seconds to perform undo.
    // after 5 seconds, 'archive' will be the winner of the race-condition
    const { undo, archive } = yield race({
      undo: take(action => action.type === 'UNDO' && action.undoId === undoId),
      archive: call(delay, 5000)
    })

    // hide undo UI element, the race condition has an answer
    needsHideUndo = false
    yield put(actions.hideUndo(undoId))

    if (archive) {
      // make the API call to apply the changes remotely
      needsRevertOptimisticUpdate = false
      yield call(updateThreadApi, thread)
    }

  } finally {
    // no need to check for cancellation, just execute this code when the 
    // try-block has finished (it will include cancellation as well. but also
    // the actual undo, error handling, ...)

    if (needsHideUndo) {
      yield put(actions.hideUndo(undoId))
    }

    if (needsRevertOptimisticUpdate) {
      // revert thread to previous state
      yield put(actions.updateThread({ id: threadId, archived: false }))
    }
  }
}

function* main() {
  while (true) {
    // wait for an ARCHIVE_THREAD to happen
    const action = yield take(`ARCHIVE_THREAD`)
    // use spawn to execute onArchive in a non-blocking fashion (which also
    // prevents cancelation when main saga gets cancelled.
    // This helps us in keeping state in sync between server and client
    yield fork(onArchive, action)
  }
}
```

But I'm not sure this is a proper solution. And it is harder to understand.

Anyway, I will update the code sample with your comments

**Comment by HansDP** [comment]
Did I updated this pull request correctly? (because i see 3 commits instead of one)
If so, what should I do? I'm not used to using github (sorry for the trouble)

**Comment by Andarist** [comment]
Only a single commit should be wanting to get merged.

You should:
`git rebase -i HEAD~3`
you will see 3 commits - use `p d s` on them in that order, so 1st will get picked, 2nd dropped, 3rd squashed
you will get asked about new commit message - just keep single `Update Undo recipe`

I aint sure if you did it right, so you might skip next steps if you have already done them:
`git remote add upstream https://github.com/yelouafi/redux-saga.git`
`git fetch upstream`

And finally
`git rebase upstream/master`
`git force --push`

**Comment by Andarist** [comment]
Thanks, real good example, leveraging `take(predicate)` which is rly neat and underpromoted in the docs 🎉

**Pull request merged**